### ISSUE TITLE
model.to(device) rather than model.to('cuda')

### ISSUE THREAD

**Issue by lsqshr** [TissueImageAnalytics/tiatoolbox#249]
### Description

Thanks for making the package.

For all the `*.predict(...)` with GPU, would it make sense to allow users to pass in the device id rather than `on_gpu=True`?

The code [here](https://github.com/TissueImageAnalytics/tiatoolbox/blob/de045ba53d25db0b15c64bfd4f595a10a2a58a5d/tiatoolbox/utils/misc.py) could pass in the device id like 'cuda:1' to model.to(device_id) instead of model.to('cuda'). It can allow parallel slide processing when more than one GPU is available.

**Comment by shaneahmed** [comment]
Thanks. It is a nice suggestion. We should be able to add this.

**Comment by lsqshr** [comment]
Awesome!

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by mostafajahanifar** [comment]
Thanks @lsqshr for the suggestion. I will look into this.

**Comment by mostafajahanifar** [comment]
Dear @lsqshr,

We have discussed this internally. Apparently, the idea has been previously pitched when we were designing the model framework of TIAToolbox, however, that would idea would have brought some complications to the design that team decided to stick to the simple form of `model.to('cuda')` instead of `model.to(device_id)`. 
For example, in the case of using mutli-GPU computation for a single task this can be problematic, it can also be confusing for the user to specify the GPUs for different experiments using this method. Not to mention that moving tensors to the desired device should be considered as well, which will make the model development confusing and error-prone in the future. Therefore, considering these technical challenges, we decided to go with the simple solution while knowing that users can easily control the GPU selection for their experiments using the `CUDA_VISIBLE_DEVICES` variable.

Either if you are using python manuscripts or our command line interface (CLI), it's easy to set the GPU(s) that you want to run your experiments with using `CUDA_VISIBLE_DEVICES` variable:
**1- Experiment 1:** In case of using python code, you just need to add these two lines of code at the top of your running code to specify the IDs of GPUs you want to use:
``` 
import os
os.environ["CUDA_VISIBLE_DEVICES"]="0,3" # means that GPUs with IDs 0 and 3 will be used in this experiment
```

**2- Experiment 2:** In case of using TIAToolbox CLI, you can similarly set the GPUs using the same variable. For example, using GPU with ID=1,4 for instance segmentation with HoVer-Net model:
```
CUDA_VISIBLE_DEVICES=1,4 tiatoolbox nucleus-instance-segment --img-input "/path/to/input.csv" --mode wsi --on-gpu True --output-path "/path/to/results/"
```

In our experience, this is the optimal (simplest) way to use the toolbox on a system with multiple GPUs. I hope you find this useful and this has answered your concerns.

**Comment by lsqshr** [comment]
Thanks for the reply. It's good to know that it was discussed before.

I was trying to see how you could parallelize the processing of multiple slides (say neuclei detection) on multi-gpu servers (say 16 gpus) for maximum GPU utilization due to the I/O bottleneck. So one GPU per slide. then in each process, I could have a segmentor by passing `f'cuda:{process_id}'` as the device name. I can of course do it with a fork or bashing it but I still suggest allow the users of the library to have some freedom of defining the computing devices according to their infrastructure. Your call.

**Comment by mostafajahanifar** [comment]
To make things clear, let me briefly describe how parallel processing works in TIAToolbox models. So, as you mentioned, in order to maximize the GPU utilization on multi-GPU servers while handling I/O bottlenecks (and also some other technical matters based on our experience working with WSIs), we decided parallelize the GPU computations on patch-level, not WSIs. In other words, TIAToolbox processes WSIs one-by-one and uses all the available resources to finish processing one slide before going to the next one. For example, if you have 1000 WSIs to process on 16 GPUs, TIAToolbox will take the first WSI and patch it up, start the GPU inference on those patches with all GPU power (all 16 GPUs if you have made them available to the running session) and then go to the next WSI to process (note that the post-processing operations and results aggregation are being done in the background). Doing this has some benefits like ensuring maximum utilization of the GPU (all the time), speeding up process of GPU-intensive models, easier handling of exceptions during the process of each WSI, making the availability of results of each processed WSI faster, etc.
Note that doing one GPU per slide (as you have in mind) will need more resources as some models (like semantic segmentation models) require high amount of memory for pre - or post-processing tasks, therefore, processing multiple slides concurrently will increase this requirement (not to mention requiring separate CPU workers for the processes related to each WSI).

So, in short, we decided to distribute data on multi-GPUs systems on patch level. However, if you still like to distribute your WSIs to be processed on different GPUs, you should be able to do so without forking the toolbox. You can easily r partition your list of WSIs into 16 sub-lists and run your inference code on each of those sub-lists while specify a separate GPU for that run (using `CUDA_VISIBLE_DEVICES` variable). But this is not recommended, as you will need 16 times more CPU and memory requirements for doing the processes.

**Comment by lsqshr** [comment]
thanks for the hints. that makes sense. closing this issue for now then.
